tests/ui/new_ret_no_self.rs

```diff
@@ -0,0 +1,76 @@
+#![feature(tool_lints)]
+
+#![warn(clippy::new_ret_no_self)]
+#![allow(dead_code, clippy::trivially_copy_pass_by_ref)]
+
+fn main(){}
+
+trait R {
+    type Item;
+}
+
+struct S;
+
+impl R for S {
+    type Item = Self;
+}
+
+impl S {
+    // should not trigger the lint
+    pub fn new() -> impl R<Item = Self> {
+        S
+    }
+}
+
+struct S2;
+
+impl R for S2 {
+    type Item = Self;
+}
+
+impl S2 {
+    // should not trigger the lint
+    pub fn new(_: String) -> impl R<Item = Self> {
+        S2
+    }
+}
+
+struct S3;
+
+impl R for S3 {
+    type Item = u32;
+}
+
+impl S3 {
+    // should trigger the lint, but currently does not
```

Comment:
You can try debug printing `ret_ty.sty`, for this case it should be `TyKind::Opaque`. If it is not, then the `return_ty` function seems to alread reveal impl trait types. If that is the case, we'll need to figure out a way to get the unadjusted signature inside the `return_ty` function

Reply:
I presume you mean the concrete type that the impl trait hides. You get it by calling `cx.tcx.type_of(def_id)`.

Using the concrete type will mess up your check in https://github.com/rust-lang-nursery/rust-clippy/pull/3253/files/3da7c1013cc9c62357261e786958c479bc5c3e09#diff-76448e132e544f8a1ff89ffa05ef9566R917 though and thus create new false positives I presume

Reply:
`cx.tcx.type_of(def_id)` returns `S`, `S2`, and `S3` as you said. I am looking for something that would return `impl R<Item = Self>` or `impl R<Item = u32>`. 

Does that exist?